pkg/networkservice/common/onidle/server_test.go

```diff
@@ -0,0 +1,185 @@
+
+	clockMock := clockmock.New()
+	ctx = clock.WithClock(ctx, clockMock)
+
+	timeout := time.Hour
+	var flag atomic.Bool
+
+	server := onidle.NewServer(ctx, func() {
+		flag.Store(true)
+	}, onidle.WithTimeout(timeout))
+
+	clockMock.Add(timeout - 1)
+	conn, err := server.Request(ctx, &networkservice.NetworkServiceRequest{})
+	require.NoError(t, err)
+	clockMock.Add(timeout)
+	require.Never(t, flag.Load, testWait, testTick)
+
+	_, err = server.Close(ctx, conn)
+	require.NoError(t, err)
+	clockMock.Add(timeout - 1)
+	require.Never(t, flag.Load, testWait, testTick)
+	clockMock.Add(1)
+	require.Eventually(t, flag.Load, testWait, testTick)
+}
+
+func TestIdleNotifier_HoldingActiveRequest(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	clockMock := clockmock.New()
+	ctx = clock.WithClock(ctx, clockMock)
+
+	timeout := time.Hour
+	var flag atomic.Bool
+
+	server := onidle.NewServer(ctx, func() {
+		flag.Store(true)
+	}, onidle.WithTimeout(timeout))
+
+	clockMock.Add(timeout - 1)
+	conn1, err := server.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: "1",
+		},
+	})
+	require.NoError(t, err)
+	conn2, err := server.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: "2",
+		},
+	})
+	require.NoError(t, err)
+
+	_, err = server.Close(ctx, conn1)
+	require.NoError(t, err)
+	clockMock.Add(timeout)
+	require.Never(t, flag.Load, testWait, testTick)
+	_, err = server.Close(ctx, conn2)
+	require.NoError(t, err)
+	clockMock.Add(timeout)
+	require.Eventually(t, flag.Load, testWait, testTick)
+}
+
+func TestIdleNotifier_FailedRequest(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	clockMock := clockmock.New()
+	ctx = clock.WithClock(ctx, clockMock)
+
+	timeout := time.Hour
+	var flag atomic.Bool
+
+	server := next.NewNetworkServiceServer(
+		onidle.NewServer(ctx, func() {
+			flag.Store(true)
+		}, onidle.WithTimeout(timeout)),
+		injecterror.NewServer(),
+	)
+
+	clockMock.Add(timeout - 1)
+	require.Never(t, flag.Load, testWait, testTick)
+
+	_, err := server.Request(ctx, &networkservice.NetworkServiceRequest{})
+	require.Error(t, err)
+	clockMock.Add(timeout - 1)
+	require.Never(t, flag.Load, testWait, testTick)
+
+	clockMock.Add(1)
+	require.Eventually(t, flag.Load, testWait, testTick)
+}
+
+func TestIdleNotifier_ContextCancel(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	clockMock := clockmock.New()
+	ctx = clock.WithClock(ctx, clockMock)
+
+	timeout := time.Hour
+	var flag atomic.Bool
+
+	_ = onidle.NewServer(ctx, func() {
+		flag.Store(true)
+	}, onidle.WithTimeout(timeout))
+
+	cancel()
+	time.Sleep(time.Millisecond * 50)
```

Comment:
Please don't use time.Sleep in tests. Consider to use `Eventually`

Reply:
`require.Eventually`/`require.Never` need a condition, but we don't have any.
The issue here is that we don't have any feedback from background thread in onidle element about whether it has received info about context cancellation or not.
Previously I used `runtime.Goshed` here, to make sure that background thread gets some CPU time, but apparently it's not enough.

Reply:
I think we can use `require.Eventually` + `goleak.Find()` to verify that goroutine exited.
This check prevents potential test failures on ci. 
Note: 50ms  could be not enough for ci.

Reply:
What would be an appropriate timeout?
My concern here is that these tests use `require.Never` a lot, so test time will be proportional to timeout.